### PR TITLE
High-performance bulk rating with BatchProcessor and Ladder

### PR DESCRIPTION
**Pull request by TinBane**
### Summary              
                                                    
- Add BatchProcessor for parallel game processing with wave-based partitioning
- Add Ladder for in-place array-backed ratings with zero-copy updates 
- Achieve 2.5–7x speedup over sequential model.rate() with exact accuracy across all 5 models

### New modules

openskill/batch.py: Parallel batch processor that partitions games into conflict-free waves (no entity appears twice per wave), enabling safe concurrent execution while preserving chronological ordering. Uses ThreadPoolExecutor on free-threaded Python (3.13t/3.14t) and ProcessPoolExecutor on GIL Python. Includes _FastRating, a minimal __slots__-only replacement for model.rating() that eliminates UUID generation and __dict__ allocation in hot paths.

openskill/ladder.py — In-place rating registry storing mu/sigma in contiguous array.array('d') buffers with RatingView flyweight objects for direct array access. Includes rate_batch() for wave-ordered bulk processing and an optional Cython fast path (_cfast.pyx, built via build_cfast.py) for 10–15% additional speedup, auto-detected at runtime.

### Performance

PlackettLuce, 3000 players, best of 3 runs:

| Dataset | Original | Ladder | Speedup |
|---|---|---|---|
| Swiss (13,500 games, 1v1) | 292 ms | 117 ms | 2.5x |
| Power-law (5,000 games, mixed teams) | 304 ms | 102 ms | 2.9x |
 

  ### Other changes

  - README updated with Ladder, rate_batch, and BatchProcessor usage examples,
  Cython build instructions, and free-threaded Python guidance
  - Fixed all mypy --strict errors in batch.py and ladder.py
  - Applied black/isort formatting across all new and modified files
  - Added speed_test.py and compare_benchmark.py benchmark scripts
  - Added fast_rating_readme.md with architecture and performance documentation

### Test plan

- [x] 841 tests (465 batch + 376 ladder) covering wave partitioning, all 5 models, sequential/parallel/pipelined modes, reproducibility, and edge cases
- [x] All 766 tests pass on Python 3.10–3.14
- [x] `tox` passes: lint, build, type, all supported Python versions
- [x] All optimised approaches produce exact results vs original baseline (within 1e-9)

### Issue(s) Resolved

None

Fixes #

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under openskill.py's MIT license.

I certify the above statement is true and correct:

### CODE REVIEW

File: benchmark/batch_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+        "model": model_name,
+        "label": label,
+        "approaches": {k: v[0] for k, v in approaches.items()},
+        "match": match_str,
+        "old_time": old_time,
+    }
+
+
+def run_benchmark():
+    print("=" * 80)
```
Comment (vivekjoshy) [inline benchmark/batch_benchmark.py]: Please use `rich` library to pretty print things. We've added it as a developer dependency anyways, so might as well use it.

File: fast_rating_readme.md
Comment (vivekjoshy) [inline fast_rating_readme.md:1]: This can be moved to README as a small feature set option and then the rest should be in a sphinx docs page.

File: build_cfast.py
Comment (vivekjoshy) [inline build_cfast.py:1]: Link this to a console script in `pyproject.toml` called `openskill build fast` maybe? We can then move this out of this root folder.

**Review by vivekjoshy — CHANGES_REQUESTED** [review]
Thanks for the PR. Great work! There is a bit of stylistic and implementation changes need before it can be merged in, but otherwise wonderful PR.

Please also bring up the test coverage to 100% or properly exclude sections impossible to test with mocks. 

Note: I haven't tested it yet locally, I'll do a proper review later.

**Comment by TinBane** [comment]
**Addressed all review feedback. Key changes:**
                                                                                
- Fixed the sysconfig string comparison bug and removed the misleading n_workers param from rate_batch()
- Consolidated the Cython build into the existing setup.py — no separate build script needed. It auto-builds when Cython is available at install time (pip install openskill[cython]), renamed the extras from "fast" to "cython", installs with cython required.
- Replaced fast_rating_readme.md with proper Sphinx docs (bulk_rating.rst) and updated the README
- Rewrote batch_benchmark.py using rich
- Added comprehensive tests for batch.py and ladder.py (98.46% coverage)
- Added Sphinx docstrings to all public API surfaces

File: benchmark/batch_benchmark.py (hunk 1 above)
Comment (TinBane) [reply benchmark/batch_benchmark.py]: We are using it now :)

File: build_cfast.py
Comment (TinBane) [reply build_cfast.py:1]: I've improved how this works, it's much more streamlined and pythonic and follows the pip norms.

**Comment by TinBane** [comment]
I've implemented all the requested fixes, got the tests to 100% coverage, and I've adjusted the way I was writing the code to match your conventions better. I'm hoping you can review it and it should be ready to pull.
